### PR TITLE
Python pypi package work

### PR DESCRIPTION
**Pull request by lindkvis**
PyPI package generation for building with manylinux2014 docker-containers in order to support most Linux-distributions (post 2014).

Split into reasonable chunks and now capable of building packages with official manylinux2014 images.

### CODE REVIEW

**Comment by lindkvis** [comment]
This should now be ready for evaluation. Please let me know if you have any issues or questions with it.

**Comment by joakim-hove** [comment]
Good thank you - just a heads up that I quote offline for a couple of days - will look at this Tuesday day next week.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

File: CMakeLists.txt [hunk 1]
```diff
@@ -284,6 +294,18 @@ if (OPM_ENABLE_PYTHON)
...
+    set(opm-common_PYTHON_LINKAGE "'-Bdynamic'")
+  else()
+    set(opm-common_PYTHON_LINKAGE "'-Bstatic', '-static-libstdc++'")
+  endif()
+
+  set(opm-common_PYTHON_PACKAGE_VERSION ${OPM_PYTHON_PACKAGE_VERSION_TAG})
+
+  # Generate versioned setup.py
+  configure_file (${PROJECT_SOURCE_DIR}/python/setup.py.in ${PROJECT_BINARY_DIR}/python/setup.py)
+  configure_file (${PROJECT_SOURCE_DIR}/python/README.md ${PROJECT_BINARY_DIR}/python/README.md COPYONLY)
```
Comment (joakim-hove) [inline CMakeLists.txt:307]: If it is purely a copy I guess the `file(COPY ...)` could be used?

**Comment by joakim-hove** [comment]
As I understand it the intention is that the pypi packaging should not be invoked from a normal cmake based build, but rather some external script? In that case maybe all the strictly pypi related files could be assembled in a `pypi/` directory?

**Comment by joakim-hove** [comment]
Not clear to me why the build fails - but that must obviously be fixed before we can consider to merge this.

File: CMakeLists.txt (hunk 1 above)
Comment (lindkvis) [reply CMakeLists.txt:307]: sure!

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
I guess the remaining jenkins test failure can be traced back to the reorganization of the code in this PR - i.e. the removal of an extra `python/` directory.

**Comment by lindkvis** [comment]
> 
> 
> I guess the remaining jenkins test failure can be traced back to the reorganization of the code in this PR - i.e. the removal of an extra `python/` directory.

That makes sense. I'm looking into them now

**Comment by lindkvis** [comment]
Da tror jeg at Python-testene skal være ok. Kan prøve å starte bygg selv, men regner med at kommandoen deres kun reagerer på folk som er registrert som admin/maintainere?

**Comment by lindkvis** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
> Kan prøve å starte bygg selv, men regner med at kommandoen deres kun reagerer på folk som er registrert som admin/maintainere?

Yes that is correct - not a very exclusive right we are talking about here, if you are to work more on opm-common you can certainly get the privelige.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
OK good; this is green now. If @lindkvis considers this complete I am ready to merge it.

**Comment by lindkvis** [comment]
Yeah. I'm happy with it. There is a secondary batch of work to be done to pass in tokens for upload to PyPI automatically, but as it stands it is at least useful and makes it possible to generate opm-common packages for PyPI via some manual commands.

**Comment by magnesj** [comment]
Thanks to all involved (@lindkvis @akva2 @joakim-hove @anders-kiaer ) to get this feature in place. Use of opm-common from Python projects will now become much easier. Thank you!

**Comment by anders-kiaer** [comment]
Thanks for the work and collaboration 👍🎉 Getting the Python bindings with `pip install opm` is a huge simplification for different research and implementation projects 👏 🚀 

I guess #1210 is closed, or very near to be closed now, since the majority of the research/work is done (only remaining is as you say @lindkvis to add the PyPI tokens and the top-level commands in a suitable place in a CI/CD workflow, such that we ensure that future releases of `opm-common` in GitHub triggers automatic release also to PyPI).

**Comment by anders-kiaer** [comment]
@lindkvis @joakim-hove @magnesj @akva2 

From OPM mailing list, I hear that OPM feature freeze is today (`2020.10.15`) and final release (if thing goes according to plan) on `2020.10.29`. When tagging `opm-common` (this repository) soon, I hope the new workflow can be used to also do a new deployment to PyPI (https://pypi.org/project/opm/)? Will it happen automatically as a CD workflow*, or does someone need to still push some buttons somewhere?

* I guess if fully automatic, there might be some issues on this first tag triggered CD PyPI deploy since there is already a release named `2020.10` (and `2020.10.1`) on PyPI? Maybe might need to call the new PyPI version `2020.10.2` (even though the corresponding git tag probably will be `2020.10` in `opm-common`)?

**Comment by anders-kiaer** [comment]
We have compiled binaries of `opm` on PyPI for Python {3.5, 3.6, 3.7, 3.8} (https://pypi.org/project/opm/#files). Python 3.9 is released now. Will automatically compiling for 3.9 be a matter of changing some lines somewhere @lindkvis? 🐍🎉

**Comment by lindkvis** [comment]
Hi.

Adding Python 3.9 requires altering python/generate-pypi-package.sh:
- [ ] Add an export line at the top of the script with the path to the python3.9 executable. This will just be export PYTHON39=/opt/python/cp39-cp39/bin/python unless the default Python build settings have changed in an incompatible way (the default used to have an "m"-postfix like cp37-cp37m) but this is unlikely to have changed now.
- [ ] Copy and alter a CMake block below, which should be fairly obvious.

There is no automatic PyPI workflow at the moment as it is not added to any CI/CD yet. Instead a developer follows the python/pypi-howto.txt from the current release branch. 
- It will be automatically tagged with the OPM version number of the branch you are on
- Test on testpypi first and if those packages work ok. Then repeat the upload to real PyPI (without the "--repository testpypi" in the command) if you are happy.
- If somehow the packages that end up at real PyPI are bad, you can repeat the process but use the line from pypi-howto.txt that adds the extra version tag, because you **can never overwrite** packages in PyPI only add new ones so you have to increase version every time you upload a change.

**Comment by lindkvis** [comment]
Adding an automated workflow will require a bit of extra work that is likely to be worth it in the future.

**Comment by anders-kiaer** [comment]
FWIW - here is an example of how `ecl` binaries are built automatically using GitHub actions: https://github.com/equinor/ecl/blob/master/.github/workflows/testing.yml, which can e.g. run in parallel with the Jenkins workflow. Secrets/credentials are stored easily using the settings tab on the repository by an admin.

Artifacts are uploaded, so if needed they can also be downloaded and tested locally (e.g. see https://github.com/equinor/ecl/actions/runs/343228077)

**Comment by berland** [comment]
Chiming in to support auto deploy of pypi packages even for release candidates (opm-common 2020.10 broke ecl2df, but now fixed, and this is nice to be able catch early).

https://github.com/equinor/ecl2df/commit/4f78d35214fe93fdf808038cb859a1aeb07ff283

**Comment by alfbr** [comment]
Just to clarify, is it sufficient if we have deployment of Pypi packages for each RC during each release of OPM?

**Comment by lindkvis** [comment]
What would make sense for you is to generate packages and upload to testpypi regularly with an incremented build number.

I suppose you could also automatically upload to the real pypi for each release tag, but that's assuming a pypi-package has tested ok from testpypi.

Gaute

**Comment by berland** [comment]
> Just to clarify, is it sufficient if we have deployment of Pypi packages for each RC during each release of OPM?

That is probably sufficient.

**Pull request merged**